src/optimize/watch/watch_cache.ts

```diff
@@ -0,0 +1,151 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { createHash } from 'crypto';
+import { existsSync, readFileSync, writeFileSync } from 'fs';
+import { resolve } from 'path';
+import { promisify } from 'util';
+
+import del from 'del';
+import deleteEmpty from 'delete-empty';
+import globby from 'globby';
+import mkdirp from 'mkdirp';
+
+const mkdirpAsync = promisify(mkdirp);
+
+interface Params {
+  log: (tags: string[], data: string) => void;
+  outputPath: string;
+  dllsPath: string;
+  cachePath: string;
+}
+
+interface WatchCacheStateContent {
+  optimizerConfigSha?: string;
+  yarnLockSha?: string;
+}
+
+export class WatchCache {
+  private readonly log: Params['log'];
+  private readonly outputPath: Params['outputPath'];
+  private readonly dllsPath: Params['dllsPath'];
+  private readonly cachePath: Params['cachePath'];
+  private readonly statePath: string;
+  private readonly cacheState: WatchCacheStateContent;
+  private diskCacheState: WatchCacheStateContent;
+
+  constructor(params: Params) {
+    this.log = params.log;
+    this.outputPath = params.outputPath;
+    this.dllsPath = params.dllsPath;
+    this.cachePath = params.cachePath;
+
+    this.statePath = resolve(this.outputPath, 'watch_optimizer_cache_state.json');
+    this.cacheState = {};
+    this.diskCacheState = this.read();
+    this.cacheState.yarnLockSha = this.buildYarnLockSha();
+    this.cacheState.optimizerConfigSha = this.buildOptimizerConfigSha();
+  }
+
+  public async tryReset() {
+    if (!this.isResetNeeded()) {
+      return;
+    }
+
+    await this.reset();
+  }
+
+  public async reset() {
+    this.log(['info', 'optimize:watch_cache'], 'The optimizer watch cache will reset');
+
+    // start by deleting the state file to lower the
+    // amount of time that another process might be able to
+    // successfully read it once we decide to delete it
+    del.sync(this.statePath);
+
+    // first delete some empty folder that left
+    // from any previous cache reset action
+    deleteEmpty.sync(`${this.cachePath}`);
+
+    // delete everything in optimize/.cache directory
+    // except ts-node
+    await del(await globby([`${this.cachePath}`, `!${this.cachePath}/ts-node/**`], { dot: true }));
+
+    // delete dlls
+    await del(this.dllsPath);
+    await mkdirpAsync(this.dllsPath);
+
+    // re-write new cache state file
+    this.write();
+
+    this.log(['info', 'optimize:watch_cache'], 'The optimizer watch cache has reset');
+  }
+
+  private buildShaWithMultipleFiles(filePaths: string[]) {
+    const shaHash = createHash('sha1');
+
+    filePaths.forEach((filePath: string) => {
+      if (existsSync(filePath)) {
+        shaHash.update(readFileSync(filePath), 'utf8');
```

Comment:
Using sync is definitely easy but have you tried using the promise versions of readFile/exists?. Curious if those would be any faster since they wouldn't block the whole process.

If you do use the async versions, note that `exists` is deprecated in favor of `stat`

Reply:
done @joshdover 😄